Migrate CacheStorage data to origin directory

szewai opened the pull request:
#### 5d9c260954a94cf6ce71bf69ab4b23c92b8e3e02
<pre>
Migrate CacheStorage data to origin directory
<a href="https://bugs.webkit.org/show_bug.cgi?id=250159">https://bugs.webkit.org/show_bug.cgi?id=250159</a>
rdar://103931715

Reviewed by Youenn Fablet.

Each partitioned origin has a directory that stores data only that origin can access. The directory is owned by
OriginStorageManager. Now that CacheStorage is managed by OriginStorageManager, we could migrate its data from
CacheStorage directory to origin directory. This change has a few benefits:
1. Current CacheStorage directory is under Caches/, which could be evicted as http cache and does not match spec
(<a href="https://w3c.github.io/ServiceWorker/#cache-lifetimes).">https://w3c.github.io/ServiceWorker/#cache-lifetimes).</a> Moving CacheStorage data to origin direcotry will make it not
purgeable.
2. Current CacheStorage directory also separates data by origin -- in CacheStorage/, each origin has a directory, and
each origin directory has an origin file. By moving data to origin directory, we no longer need the duplicate origin
file and directory under CacheStorage/.
3. Client does not need to set a separate CacheStorageDirectory, WebKit would use GeneralStorageDirectory (where origin
directories are located) for different types.

When we did this (migrating data from custom directory to unified origin directory) for LocalStorage and IndexedDB,
we used a boolean shouldUseCustomStoragePaths to indicate whether the data should be put under origin directory. We
cannot reuse this binary-state flag, since there will be three states:
1. All types use custom directory (non-Cocoa platforms)
2. IndexdDB and LocalStorage use origin directory (Cocoa platforms nowadays)
3. CacheStorage, IndexdDB and LocalStorage use origin directory (Cocoa platforms after the change is shipped)
Therefore, this patch adds UnifiedOriginStorageLevel to replace shouldUseCustomStoragePaths, which can represent more
states.

Test: WKWebsiteDataStore.MigrateCacheStorageDataToGeneralStorageDirectory

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::CacheStorageManager::cacheStorageOriginDirectory):
(WebKit::CacheStorageManager::copySaltFileToOriginDirectory):
(WebKit::CacheStorageManager::saltFilePath const):
(WebKit::CacheStorageManager::CacheStorageManager):
(WebKit::CacheStorageManager::cacheStorageSalt): Deleted.
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::originStorageManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::StorageBucket):
(WebKit::OriginStorageManager::StorageBucket::cacheStorageManager):
(WebKit::OriginStorageManager::StorageBucket::isEmpty):
(WebKit::OriginStorageManager::StorageBucket::fetchDataTypesInListFromDisk):
(WebKit::OriginStorageManager::StorageBucket::deleteCacheStorageData):
(WebKit::OriginStorageManager::StorageBucket::deleteEmptyDirectory):
(WebKit::OriginStorageManager::StorageBucket::resolvedLocalStoragePath):
(WebKit::OriginStorageManager::StorageBucket::resolvedIDBStoragePath):
(WebKit::OriginStorageManager::StorageBucket::resolvedCacheStoragePath):
(WebKit::OriginStorageManager::StorageBucket::resolvedPath):
(WebKit::OriginStorageManager::OriginStorageManager):
(WebKit::OriginStorageManager::defaultBucket):
(WebKit::OriginStorageManager::quotaManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/Shared/WebsiteData/UnifiedOriginStorageLevel.h: Added.
(WebKit::convertToUnifiedOriginStorageLevel):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(toWKUnifiedOriginStorageLevel):
(toUnifiedOriginStorageLevel):
(-[_WKWebsiteDataStoreConfiguration unifiedOriginStorageLevel]):
(-[_WKWebsiteDataStoreConfiguration setUnifiedOriginStorageLevel:]):
(-[_WKWebsiteDataStoreConfiguration shouldUseCustomStoragePaths]): Deleted.
(-[_WKWebsiteDataStoreConfiguration setShouldUseCustomStoragePaths:]): Deleted.
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel):
(WebKit::internalFeatureEnabled): Deleted.
(WebKit::WebsiteDataStore::defaultShouldUseCustomStoragePaths): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel):
(WebKit::WebsiteDataStore::defaultShouldUseCustomStoragePaths): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::WebsiteDataStoreConfiguration):
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::unifiedOriginStorageLevel const):
(WebKit::WebsiteDataStoreConfiguration::setUnifiedOriginStorageLevel):
(WebKit::WebsiteDataStoreConfiguration::shouldUseCustomStoragePaths const): Deleted.
(WebKit::WebsiteDataStoreConfiguration::setShouldUseCustomStoragePaths): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBFileName.mm:
(runTest):
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBPersistence.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LocalStoragePersistence.mm:
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(runWebsiteDataStoreCustomPaths):
(TEST):

Canonical link: <a href="https://commits.webkit.org/258591@main">https://commits.webkit.org/258591@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b8e8d9b3e32178fecdc022b60ce1a065f599a36

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102370 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35434 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111675 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171849 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106344 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12507 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2426 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/109379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108151 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9572 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37294 "layout-tests (failure)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24326 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5009 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25749 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5154 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2187 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11182 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45242 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5896 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6900 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
```diff
@@ -433,26 +439,33 @@ void OriginStorageManager::StorageBucket::moveData(OptionSet<WebsiteDataType> ty
+        FileSystem::deleteEmptyDirectory(typeStoragePath(StorageType::IndexedDB));
+        FileSystem::deleteEmptyDirectory(m_customCacheStoragePath);
+        break;
+    case UnifiedOriginStorageLevel::ServiceWorker:
```
ServiceWorker is not very descriptive. I guess All would be better for now.

szewai replied:
I picked `ServiceWorker` as I am not sure whether we are going to migrate more types later. Maybe we could go with "Primary" or "Standard" to indicate these are primary storage types.

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
```diff
@@ -1254,7 +1254,7 @@ HTTPServer server([&] (Connection connection) {
     auto websiteDataStoreConfiguration = adoptNS([[_WKWebsiteDataStoreConfiguration alloc] init]);
     websiteDataStoreConfiguration.get().generalStorageDirectory = generalStorageDirectory;
-    websiteDataStoreConfiguration.get().shouldUseCustomStoragePaths = false;
+    websiteDataStoreConfiguration.get().unifiedOriginStorageLevel = _WKUnifiedOriginStorageLevelBasic;
```
Would it make also sense to test _WKUnifiedOriginStorageLevelServiceWorker?

szewai replied:
I don't think we need to change the old tests since they only aim to check migration of IndexedDB and LocalStorage data. We could add a test to verify different behaviors on different values of _WKUnifiedOriginStorageLevelBasic  if necessary.

szewai committed:
Migrate CacheStorage data to origin directory
https://bugs.webkit.org/show_bug.cgi?id=250159
rdar://103931715

Reviewed by Youenn Fablet.

Each partitioned origin has a directory that stores data only that origin can access. The directory is owned by
OriginStorageManager. Now that CacheStorage is managed by OriginStorageManager, we could migrate its data from
CacheStorage directory to origin directory. This change has a few benefits:
1. Current CacheStorage directory is under Caches/, which could be evicted as http cache and does not match spec
(https://w3c.github.io/ServiceWorker/#cache-lifetimes). Moving CacheStorage data to origin direcotry will make it not
purgeable.
2. Current CacheStorage directory also separates data by origin -- in CacheStorage/, each origin has a directory, and
each origin directory has an origin file. By moving data to origin directory, we no longer need the duplicate origin
file and directory under CacheStorage/.
3. Client does not need to set a separate CacheStorageDirectory, WebKit would use GeneralStorageDirectory (where origin
directories are located) for different types.

When we did this (migrating data from custom directory to unified origin directory) for LocalStorage and IndexedDB,
we used a boolean shouldUseCustomStoragePaths to indicate whether the data should be put under origin directory. We
cannot reuse this binary-state flag, since there will be three states:
1. All types use custom directory (non-Cocoa platforms)
2. IndexdDB and LocalStorage use origin directory (Cocoa platforms nowadays)
3. CacheStorage, IndexdDB and LocalStorage use origin directory (Cocoa platforms after the change is shipped)
Therefore, this patch adds UnifiedOriginStorageLevel to replace shouldUseCustomStoragePaths, which can represent more
states.

Test: WKWebsiteDataStore.MigrateCacheStorageDataToGeneralStorageDirectory

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::CacheStorageManager::cacheStorageOriginDirectory):
(WebKit::CacheStorageManager::copySaltFileToOriginDirectory):
(WebKit::CacheStorageManager::saltFilePath const):
(WebKit::CacheStorageManager::CacheStorageManager):
(WebKit::CacheStorageManager::cacheStorageSalt): Deleted.
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::originStorageManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::StorageBucket):
(WebKit::OriginStorageManager::StorageBucket::cacheStorageManager):
(WebKit::OriginStorageManager::StorageBucket::isEmpty):
(WebKit::OriginStorageManager::StorageBucket::fetchDataTypesInListFromDisk):
(WebKit::OriginStorageManager::StorageBucket::deleteCacheStorageData):
(WebKit::OriginStorageManager::StorageBucket::deleteEmptyDirectory):
(WebKit::OriginStorageManager::StorageBucket::resolvedLocalStoragePath):
(WebKit::OriginStorageManager::StorageBucket::resolvedIDBStoragePath):
(WebKit::OriginStorageManager::StorageBucket::resolvedCacheStoragePath):
(WebKit::OriginStorageManager::StorageBucket::resolvedPath):
(WebKit::OriginStorageManager::OriginStorageManager):
(WebKit::OriginStorageManager::defaultBucket):
(WebKit::OriginStorageManager::quotaManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/Shared/WebsiteData/UnifiedOriginStorageLevel.h: Added.
(WebKit::convertToUnifiedOriginStorageLevel):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(toWKUnifiedOriginStorageLevel):
(toUnifiedOriginStorageLevel):
(-[_WKWebsiteDataStoreConfiguration unifiedOriginStorageLevel]):
(-[_WKWebsiteDataStoreConfiguration setUnifiedOriginStorageLevel:]):
(-[_WKWebsiteDataStoreConfiguration shouldUseCustomStoragePaths]): Deleted.
(-[_WKWebsiteDataStoreConfiguration setShouldUseCustomStoragePaths:]): Deleted.
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel):
(WebKit::internalFeatureEnabled): Deleted.
(WebKit::WebsiteDataStore::defaultShouldUseCustomStoragePaths): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel):
(WebKit::WebsiteDataStore::defaultShouldUseCustomStoragePaths): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::WebsiteDataStoreConfiguration):
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::unifiedOriginStorageLevel const):
(WebKit::WebsiteDataStoreConfiguration::setUnifiedOriginStorageLevel):
(WebKit::WebsiteDataStoreConfiguration::shouldUseCustomStoragePaths const): Deleted.
(WebKit::WebsiteDataStoreConfiguration::setShouldUseCustomStoragePaths): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBFileName.mm:
(runTest):
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBPersistence.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LocalStoragePersistence.mm:
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(runWebsiteDataStoreCustomPaths):
(TEST):

Canonical link: https://commits.webkit.org/258591@main
